components/protocol-wasm/src/lib.rs

```diff
@@ -0,0 +1,213 @@
+          for (key, value) in &package.state.permissions {
+              permissions.push(protocol::PermissionEntry {
+                  key_id: key.to_string(),
+                  permissions: value
+                      .into_iter()
+                      .map(|p: &package::model::Permission| perm_binding(p))
+                      .collect(),
+              })
+          }
+          for (key, value) in &package.state.releases {
+            let t: DateTime<Utc> = value.timestamp.into();
+            releases.push(protocol::Release {
+              version: key.to_string(),
+              by: value.by.to_string(),
+              timestamp: t.to_rfc3339(),
+              state: match &value.state {
+                package::ReleaseState::Released{ content } => protocol::ReleaseState::Released(protocol::Released {
+                  content: protocol::DynHash {
+                    algo: protocol::HashAlgorithm::Sha256,
+                    bytes: content.bytes().to_vec()
+                  }
+                }),
+                package::ReleaseState::Yanked{ by, timestamp } => {
+                  let ts: DateTime<Utc> = (*timestamp).into();
+                  protocol::ReleaseState::Yanked(protocol::Yanked {
+                    by: by.to_string(),
+                    timestamp: ts.to_string()
+                  })
+                }
+              }
+            })
+          }
+          for (key, value) in &package.state.keys {
+              keys.push(protocol::KeyEntry {
+                  key_id: key.to_string(),
+                  public_key: value.to_string(),
+              })
+          }
+        }
+        if let Some(head) = package.state.head() {
+          heads.push(protocol::LogLeaf {
+              log_id: LogId::package_log::<Sha256>("funny").to_string(),
+              record_id: head.digest.clone().to_string(),
+          });
+        } 
+        return protocol::PackageInfo {
+            name: package.name,
+            checkpoint: package.checkpoint,
+            state: protocol::Validator {
+                algorithm: Some(protocol::HashAlgorithm::Sha256),
+                head: Some(protocol::Head {
+                    digest: protocol::RecordId::DynHash(protocol::DynHash {
+                        algo: protocol::HashAlgorithm::Sha256,
+                        bytes: package
+                            .state
+                            .head
+                            .as_ref()
+                            .map(|h| h.digest.0.bytes().to_vec()).unwrap(),
+                    }),
+                    timestamp: package.state.head.map(|h| {
+                        let t: DateTime<Utc> = h.timestamp.into();
+                        t.to_rfc3339()
+                    }),
+                }),
+                permissions,
+                releases,
+                keys: Some(keys),
+            },
+            heads
+        };
+    }
+}
+
+bindings::export!(Component);
```

Comment:
Is putting the bindings gen here a convention? I'd expect it to be as early in the file as rustfmt will allow.

Reply:
Moved up to just after all the use statements, could probably move it to the line 1 too if you like